Extension/src/LanguageServer/copilotCompletionContextProvider.ts

```diff
@@ -0,0 +1,154 @@
+const completionContextCache: Map<string, CodeSnippet[]> = new Map<string, CodeSnippet[]>();
+const cppDocumentSelector: DocumentSelector = [{ language: 'cpp' }, { language: 'c' }];
+
+class DefaultValueFallback extends Error {
+    static readonly DefaultValue = "DefaultValue";
+    constructor() { super(DefaultValueFallback.DefaultValue); }
+}
+
+class CancellationError extends Error {
+    static readonly Cancelled = "Cancelled";
+    constructor() { super(CancellationError.Cancelled); }
+}
+
+let completionContextCancellation = new vscode.CancellationTokenSource();
+
+// Mutually exclusive values for the kind of snippets. They either come from the cache,
+// are computed, or the computation is taking too long and no cache is present. In the latter
+// case, the cache is computed anyway while unblocking the execution flow returning undefined.
+enum SnippetsKind {
+    Computed = 'computed',
+    CacheHit = 'cacheHit',
+    CacheMiss = 'cacheMiss'
+}
+
+// Get the default value if the timeout expires, but throws an exception if the token is cancelled.
+async function waitForCompletionWithTimeoutAndCancellation<T>(promise: Promise<T>, defaultValue: T | undefined,
+    timeout: number, token: vscode.CancellationToken): Promise<[T | undefined, SnippetsKind]> {
+    const defaultValuePromise = new Promise<T>((resolve, reject) => setTimeout(() => {
+        if (token.isCancellationRequested) {
+            reject('DefaultValuePromise was cancelled');
+        } else {
+            reject(new DefaultValueFallback());
+        }
+    }, timeout));
+    const cancellationPromise = new Promise<T>((_, reject) => {
+        token.onCancellationRequested(() => {
+            reject(new CancellationError());
+        });
+    });
+    let snippetsOrNothing: T | undefined;
+    try {
+        snippetsOrNothing = await Promise.race([promise, cancellationPromise, defaultValuePromise]);
+    } catch (e) {
+        if (e instanceof DefaultValueFallback) {
+            return [defaultValue, defaultValue !== undefined ? SnippetsKind.CacheHit : SnippetsKind.CacheMiss];
+        }
+
+        // Rethrow the error for cancellation cases.
+        throw e;
+    }
+
+    return [snippetsOrNothing, SnippetsKind.Computed];
+}
+
+// Get the completion context with a timeout and a cancellation token.
+// The cancellationToken indicates that the value should not be returned nor cached.
+async function getCompletionContextWithCancellation(documentUri: string, caretOffset: number,
+    startTime: number, out: Logger, token: vscode.CancellationToken): Promise<CodeSnippet[]> {
+    try {
+        const activeEditor: vscode.TextEditor | undefined = vscode.window.activeTextEditor;
+        if (!activeEditor ||
+            activeEditor.document.uri.toString() !== vscode.Uri.parse(documentUri).toString()) {
+            return [];
+        }
+
+        const snippets = await clients.ActiveClient.getCompletionContext(activeEditor.document.uri, caretOffset, token);

```

Comment:
I think this should call `clients.getClientFor(documentUri)` instead of clients.ActiveClient.